tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go

```diff
@@ -0,0 +1,104 @@
+package declarationcommentmismatch
+
+import (
+	"fmt"
+	"go/ast"
+	"go/token"
+	"strings"
+	"unicode"
+
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/passes/inspect"
+	"golang.org/x/tools/go/ast/inspector"
+)
+
+const doc = `check that there are no functions, interfaces or structs preceded by descriptive comments of mismatching capitalization`
+
+// Analyzer is the analyzer.
+var Analyzer = &analysis.Analyzer{
+	Name:     "declarationcommentmismatch",
+	Doc:      doc,
+	Requires: []*analysis.Analyzer{inspect.Analyzer},
+	Run:      run,
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	inspectResult := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector)
+	nodeFilter := []ast.Node{
+		(*ast.File)(nil),
+		(*ast.FuncDecl)(nil),
+		(*ast.ValueSpec)(nil),
+		(*ast.TypeSpec)(nil),
+		(*ast.GenDecl)(nil),
+	}
+
+	inspectResult.Nodes(nodeFilter, func(n ast.Node, push bool) bool {
+		if !push {
+			return true
+		}
+
+		if astFile, ok := n.(*ast.File); ok {
+			file := pass.Fset.File(astFile.Pos())
+			if file != nil && strings.HasSuffix(file.Name(), "_test.go") {
+				return false
+			}
+			return true
+		}
+
+		if astFunction, ok := n.(*ast.FuncDecl); ok {
+			checkCommentCaseMatches(pass, astFunction.Doc, astFunction.Name.String(), "function", astFunction.Pos())
+			return true
+		}
+
+		if astVar, ok := n.(*ast.ValueSpec); ok {
+			checkCommentCaseMatches(pass, astVar.Doc, astVar.Names[0].String(), "variable", astVar.Pos())
+			return true
+		}
+
+		if astGenDecl, ok := n.(*ast.GenDecl); ok {
+			documentation := astGenDecl.Doc
+			for _, spec := range astGenDecl.Specs {
+				if astType, ok2 := spec.(*ast.TypeSpec); ok2 {
+					switch astType.Type.(type) {
+					case *ast.ArrayType:
+						checkCommentCaseMatches(pass, documentation, astType.Name.String(), "array", astType.Pos())
+					case *ast.StructType:
+						checkCommentCaseMatches(pass, documentation, astType.Name.String(), "struct", astType.Pos())
+					case *ast.InterfaceType:
+						checkCommentCaseMatches(pass, documentation, astType.Name.String(), "interface", astType.Pos())
+					case *ast.MapType:
+						checkCommentCaseMatches(pass, documentation, astType.Name.String(), "map", astType.Pos())
+					case *ast.ChanType:
+						checkCommentCaseMatches(pass, documentation, astType.Name.String(), "channel", astType.Pos())
+					}
+					return true
+				}
+			}
+		}
+
+		return true
+	})
+	return nil, nil
+}
+
+func checkCommentCaseMatches(pass *analysis.Pass, doc *ast.CommentGroup, objectName string, objectType string, position token.Pos) {
+	if doc == nil {
+		return
+	}
+	commentSplit := strings.Split(doc.List[0].Text, " ")
+	if len(commentSplit) < 2 {
+		return
+	}
+
+	firstObjectLetter := []rune(objectName[0:1])[0]
```

Comment:
Let's be on the safe side and maybe check the length before accessing the first element.

Reply:
I can do that but technically it's unnecessary since the AST wouldn't parse it as one of the objects I'm filtering for if it didn't have a name of at least length 1.

Reply:
I do not know the precise behavior of the parser, but I know that there are anonymous functions in go and they might be a corner case. As we do not have any tests for that, I would suggest to be on the safe side.